heron/packing/src/java/com/twitter/heron/packing/PackingUtils.java

```diff
@@ -219,4 +222,142 @@ public static int allocateNewContainer(ArrayList<Container> containers, Resource
     return componentsToScaleUp;
   }
 
+  /**
+   * Identifies the resources reclaimed by the components that will be scaled down
+   *
+   * @return Total resources reclaimed
+   */
+  public static Resource getScaleDownResource(TopologyAPI.Topology topology, Map<String, Integer>
+      componentChanges, Resource defaultInstanceResources) {
+    double cpu = 0;
+    long ram = 0;
+    long disk = 0;
+    Map<String, Long> ramMap = TopologyUtils.getComponentRamMapConfig(topology);
+    Map<String, Integer> componentsToScaleDown = PackingUtils.getComponentsToScaleDown(
+        componentChanges);
+    for (String component : componentsToScaleDown.keySet()) {
+      int parallelismChange = -componentChanges.get(component);
+      cpu += parallelismChange * defaultInstanceResources.getCpu();
+      disk += parallelismChange * defaultInstanceResources.getDisk();
+      if (ramMap.containsKey(component)) {
+        ram += parallelismChange * ramMap.get(component);
+      } else {
+        ram += parallelismChange * defaultInstanceResources.getRam();
+      }
+    }
+    return new Resource(cpu, ram, disk);
+  }
+
+  /**
+   * Identifies the resources needed for the components that will be scaled up
+   *
+   * @return Total resources needed
+   */
+  public static Resource getScaleUpResource(TopologyAPI.Topology topology, Map<String, Integer>
```

Comment:
There's a lot of duplicate logic in `getScale[Up|Down]Resource` and `getComponentsToScale[Up|Down]` that could be replaced by refactoring into generic methods that take an enum that represents whether we're scaling up or down:
```
  enum ScalingDirection {
    UP,
    DOWN;
    
    boolean includes(int parallelismChange) {
      switch (this) {
        case UP:
          return parallelismChange > 0;
        case DOWN:
          return parallelismChange < 0;
        default:
          throw new IllegalArgumentException("...");
      }
    }
  }
```
Also, I think we could name this method more clearly. Maybe something like this maybe?
```
Resource computeTotalResourceChange(
  TopologyAPI.Topology topology,
  Map<String, Integer> componentChanges,
  Resource defaultInstanceResources,
  ScalingDirection scalingDirection);
```